src/core/include/mp-units/bits/fixed_point.h

```diff
@@ -0,0 +1,309 @@
+      // unfortunately, wide division is hard: https://en.wikipedia.org/wiki/Division_algorithm.
+      // Here, we just provide a somewhat naive implementation of long division.
+      Tl rem_hi = lhs.hi % rhs;
+      Tl rem_lo = lhs.lo;
+      Tl res_lo = 0;
+      for (std::size_t i = 0; i < base_width; ++i) {
+        // shift in one bit
+        rem_hi = (rem_hi << 1u) | (rem_lo >> (base_width - 1));
+        rem_lo <<= 1u;
+        res_lo <<= 1u;
+        // perform one bit of long division
+        if (rem_hi >= rhs) {
+          rem_hi -= rhs;
+          res_lo |= 1u;
+        }
+      }
+      return ret_t{res_hi, res_lo};
+    }
+  }
+
+  template<std::integral Rhs>
+    requires(std::numeric_limits<Rhs>::digits <= base_width)
+  friend constexpr double_width_int operator+(const double_width_int& lhs, Rhs rhs)
+  {
+    // this follows the usual (but somewhat dangerous) rules in C++; we "win", as we are the larger type.
+    // -> signed computation only of both types are signed
+    if constexpr (is_signed, std::is_signed_v<Rhs>) {
+      if (rhs < 0) return lhs - static_cast<std::make_unsigned_t<Rhs>>(-rhs);
+    }
+    Tl ret = lhs.lo + static_cast<Tl>(rhs);
+    return {lhs.hi + Th{ret < lhs.lo ? 1 : 0}, ret};
+  }
+  template<std::integral Lhs>
+  friend constexpr double_width_int operator+(Lhs lhs, const double_width_int& rhs)
+  {
+    return rhs + lhs;
+  }
+  template<std::integral Rhs>
+    requires(std::numeric_limits<Rhs>::digits <= base_width)
+  friend constexpr double_width_int operator-(const double_width_int& lhs, Rhs rhs)
+  {
+    // this follows the usual (but somewhat dangerous) rules in C++; we "win", as we are the larger type.
+    // -> signed computation only of both types are signed
+    if constexpr (is_signed, std::is_signed_v<Rhs>) {
+      if (rhs < 0) return lhs + static_cast<std::make_unsigned_t<Rhs>>(-rhs);
+    }
+    Tl ret = lhs.lo - static_cast<Tl>(rhs);
+    return {lhs.hi - Th{ret > lhs.lo ? 1 : 0}, ret};
+  }
+  template<std::integral Lhs>
+  friend constexpr double_width_int operator-(Lhs lhs, const double_width_int& rhs)
+  {
+    return rhs + lhs;
+  }
+
+  constexpr double_width_int operator-() const { return {(lo > 0 ? -1 : 0) - hi, -lo}; }
+
+  constexpr double_width_int operator>>(unsigned n) const
+  {
+    if (n >= base_width) {
+      return {static_cast<Th>(hi < 0 ? -1 : 0), static_cast<Tl>(hi >> (n - base_width))};
+    }
+    return {hi >> n, (static_cast<Tl>(hi) << (base_width - n)) | (lo >> n)};
+  }
+  constexpr double_width_int operator<<(unsigned n) const
+  {
+    if (n >= base_width) {
+      return {static_cast<Th>(lo << (n - base_width)), 0};
+    }
+    return {(hi << n) + static_cast<Th>(lo >> (base_width - n)), lo << n};
+  }
+
+  static constexpr double_width_int max() { return {std::numeric_limits<Th>::max(), std::numeric_limits<Tl>::max()}; }
+
+  Th hi;
+  Tl lo;
+};
+
+#if defined(__SIZEOF_INT128__)
+MP_UNITS_DIAGNOSTIC_PUSH
+MP_UNITS_DIAGNOSTIC_IGNORE("-Wpedantic")
```

Comment:
Something like `MP_UNITS_DIAGNOSTIC_IGNORE_PEDANTIC` would probably be a better solution here.